ccvm_simulators/solvers/dl_solver.py

```diff
@@ -114,7 +114,9 @@ def parameter_key(self, parameters):
         self._parameter_key = parameters
         self._is_tuned = False
 
-    def _calculate_drift_boxqp(self, c, s, pump, rate, feedback_scale=100, S=1):
+    def _calculate_drift_boxqp(
```

Comment:
@FarhadK-1QBit 
This is from the issue description:
```
The functions _calculate_grads_boxqp and _calculate_drift_boxqp also needs to be updated to something like this:
c_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", c * (upper_limit - lower_limit) / S + (upper_limit + lower_limit), self.q_matrix) * (upper_limit - lower_limit)/ S
c_grad_3 = self.v_vector * (upper_limit - lower_limit) / (2 * S)
```

But I am not confident how to apply this to `_calculate_grads_boxqp`, can you please help me with that?

Also, do the s_grad variables need to be updated to account for upper and lower bounds, too?

Reply:
Similarly, can we discuss the implementation for the other solvers' functions?

Reply:
For `_calculate_grads_boxqp` exactly the same happens, that is, in this line: https://github.com/1QB-Information-Technologies/ccvm/blob/b0b2c29ad28f1dae4fb385b89f9468c349548dac/ccvm_simulators/solvers/dl_solver.py#L180

Replace the code with this:
```
c_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", c * (upper_limit - lower_limit) / S + (upper_limit + lower_limit), self.q_matrix) * (upper_limit - lower_limit)/ S
c_grad_3 = self.v_vector * (upper_limit - lower_limit) / (2 * S)
```

For both `_calculate_drift_boxqp` and `_calculate_grads_boxqp` we can write similarly for the `s_grad`:
```
s_grad_1 = 0.25 * torch.einsum("bi,ij -> bj", s * (upper_limit - lower_limit) / S + (upper_limit + lower_limit), self.q_matrix) * (upper_limit - lower_limit)/ S
s_grad_3 = self.v_vector * (upper_limit - lower_limit) / (2 * S)
```

In the MF-CCVM, you can use this for both `_calculate_drift_boxqp` and `_calculate_grads_boxqp` here: 
https://github.com/1QB-Information-Technologies/ccvm/blob/b0b2c29ad28f1dae4fb385b89f9468c349548dac/ccvm_simulators/solvers/mf_solver.py#L201C1-L207C1

```
mu_term2_1 = ( -(1 / 4) * (torch.einsum("bi,ij -> bj", mu_tilde * (upper_limit - lower_limit) / S  + (upper_limit + lower_limit), self.q_matrix)) * (upper_limit - lower_limit) / S )
mu_term2_2 = -self.v_vector * (upper_limit - lower_limit) / (2 * S)
```

For the Langevin solver you can write again for both `_calculate_drift_boxqp` and `_calculate_grads_boxqp`:
```
c_drift_1 = torch.einsum("bi,ij -> bj", c *(upper_limit - lower_limit)/ (2 * S) + (upper_limit + lower_limit)/ 2, self.q_matrix)
c_drift_2 = self.v_vector

c_drift = -(c_drift_1 + c_drift_2) * (upper_limit - lower_limit) / (2 * S)
```

And for pumped-Langevin you can write for `_calculate_grads_boxqp` here: 
https://github.com/1QB-Information-Technologies/ccvm/blob/b0b2c29ad28f1dae4fb385b89f9468c349548dac/ccvm_simulators/solvers/pumped_langevin_solver.py#L129C1-L132C43

```
c_grad_1 = torch.einsum("bi,ij -> bj", c * (upper_limit - lower_limit) / (2 * S) + (upper_limit + lower_limit) / 2, self.q_matrix) * (upper_limit - lower_limit) / (2 * S)
c_grad_2 = self.v_vector * (upper_limit - lower_limit) / (2 * S)
```
The `_calculate_drift_boxqp` doesn't need to be changed here. 

Sorry for not clarifying this before. Let me know if you're unsure about any parts. I will make sure to review these updates.

Thanks so much!

Reply:
Thank you for providing all of these implementations! I appreciate that!